src/batch.rs

```diff
@@ -0,0 +1,61 @@
+//! # Batch
+//!
+//! This module contains types for batch operations. A batch operation is a
+//! collection of write operations (puts and/or deletes) that are applied
+//! atomically to the database.
+
+/// A batch of write operations (puts and/or deletes). All operations in the
+/// batch are applied atomically to the database. If multiple operations appear
+/// for a a single key, the last operation will be applied. The others will be
+/// dropped.
+///
+/// # Examples
+/// ```rust,no_run,compile_fail
+/// use slatedb::batch::{WriteBatch, WriteOp};
+///
+/// let object_store = Arc::new(LocalFileSystem::new());
+/// let db = Db::open("path/to/db".into(), object_store).await;
+///
+/// let mut batch = WriteBatch::new();
+/// batch.put(b"key1", b"value1");
+/// batch.put(b"key2", b"value2");
+/// batch.delete(b"key3");
+///
+/// db.write(batch).await;
+/// ```
+///
+/// Note that the `WriteBatch` has an unlimited size. This means that batch
+/// writes can exceed `l0_sst_size_bytes` (when `WAL` is disabled). It also
+/// means that WAL SSTs could get large if there's a large batch write.
+pub struct WriteBatch {
+    pub(crate) ops: Vec<WriteOp>,
+}
+
+impl Default for WriteBatch {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub enum WriteOp {
+    Put(Vec<u8>, Vec<u8>),
```

Comment:
Maybe WriteOp<T: AsRef<[u8]>> is better?

Reply:
``` rust
pub enum WriteBatchEntry<T: AsRef<[u8]>> {
    Put(T, T),
    Del(T)
}
```
Probably something like this, which is more general and can also avoid the related conversion of &[u8] to Vec.

Reply:
The annoying thing about the the above approach is that, in this code:

```rust
        let mut batch = WriteBatch::new();
        batch.put(b"key1", b"value1");
        batch.put(b"key2", b"value2");
```

The rust compiler will pick the T to be `&[u8; 4]` because it first sees `b"key`"`. Then it will complain that `b"value2"` is length 6 not 4. 😞  You have to then do `WriteBatch::<&[u8]>::new();`, which is kind of annoying... 🤔

Reply:
I updated the code here: https://github.com/slatedb/slatedb/pull/264/commits/cb13d84c5f34d77a86bd870838a410e0c60d30aa

I've chosen to do `&'a [u8]` rather than the approach you've taken. It's more aligned with our existing APIs and less invasive (doesn't require the `::<[u8]>` I mentioned above). I acknowledge it's less flexible. I wonder if we should change all of our APIs (put/delete/get) to use typed AsRef's, but we should have that discussion separately.